frontend/src/utils/webauthn.ts

```diff
@@ -0,0 +1,128 @@
+// Copyright 2025 New Vector Ltd.
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+// Please see LICENSE in the repository root for full details.
+
+// Polyfills for fromJSON and toJSON utils which aren't stable yet
+if (typeof window !== "undefined" && window.PublicKeyCredential) {
+  const b64urlDecode = (b64: string) =>
+    Uint8Array.from(
+      atob(b64.replace(/-/g, "+").replace(/_/g, "/")),
+      (c) => c.codePointAt(0) as number,
+    );
+  const b64urlEncode = (buf: ArrayBuffer) =>
+    btoa(
+      Array.from(new Uint8Array(buf), (b) => String.fromCodePoint(b)).join(""),
+    )
+      .replace(/\+/g, "-")
+      .replace(/\//g, "_")
+      .replace(/=+$/, "");
+
+  // TS doesn't know AuthenticatorAttestationResponseJSON and AuthenticatorAssertionResponseJSON yet
+  type AuthenticatorAttestationResponseJSON = {
+    clientDataJSON: Base64URLString;
+    authenticatorData: Base64URLString;
+    transports: string[];
+    publicKey?: Base64URLString;
+    publicKeyAlgorithm: COSEAlgorithmIdentifier;
+    attestationObject: Base64URLString;
+  };
+
+  type AuthenticatorAssertionResponseJSON = {
+    clientDataJSON: Base64URLString;
+    authenticatorData: Base64URLString;
+    signature: Base64URLString;
+    userHandle?: Base64URLString;
+  };
+
+  if (!window.PublicKeyCredential.parseCreationOptionsFromJSON) {
```

Comment:
Looks like this polyfill exists to do that: https://github.com/MasterKale/webauthn-polyfills

Reply:
I did use that at first, but it turned out that it didn't follow the spec and was missing fields.